framework/src/org/checkerframework/common/basetype/BaseTypeVisitor.java

```diff
@@ -1822,19 +1823,100 @@ protected void commonAssignmentCheck(
 
         // Use an error key only if it's overridden by a checker.
         if (!success) {
-            String valueTypeString = valueType.toString();
-            String varTypeString = varType.toString();
-            // If both types as strings are the same, try printing verbosely.
-            if (valueTypeString.equals(varTypeString)
-                    // or if neither string contains an annotation
-                    || (!valueTypeString.contains("@") && !varTypeString.contains("@"))) {
+            String valueTypeString;
+            String varTypeString;
+            if (shouldPrintVerbose(varType, valueType)) {
                 valueTypeString = valueType.toString(true);
                 varTypeString = varType.toString(true);
+            } else {
+                valueTypeString = valueType.toString();
+                varTypeString = varType.toString();
             }
             checker.report(Result.failure(errorKey, valueTypeString, varTypeString), valueTree);
         }
     }
 
+    /**
+     * Return whether or not the verbose toString should be used when printing the two annotated
+     * types.
+     *
+     * <p>If any simple toString of any annotated type in the first atm is equal to any toString of
+     * any annotated type in the second atm and if the verbose toStrings differ, then the verbose
+     * toString should be used.
+     *
+     * @param atm1 the first atm
+     * @param atm2 the second atm
+     * @return whether or not the verbose toString should be used when printing the two annotated
+     *     types.
+     */
+    private boolean shouldPrintVerbose(AnnotatedTypeMirror atm1, AnnotatedTypeMirror atm2) {
+        String atm1ToString = atm1.toString();
+        String atm2ToString = atm2.toString();
+        // If both types as strings are the same, use verbose toString.
+        if (atm2ToString.equals(atm1ToString)
+                // or if neither string contains an annotation
+                || (!atm2ToString.contains("@") && !atm1ToString.contains("@"))) {
+            return true;
+        }
+
+        // Collect the "toString" of every AnnotatedTypeMirror in atm2.
+        SimpleAnnotatedTypeScanner<Map<String, List<AnnotatedTypeMirror>>, Void> collectToStrings =
+                new SimpleAnnotatedTypeScanner<Map<String, List<AnnotatedTypeMirror>>, Void>() {
+                    Map<String, List<AnnotatedTypeMirror>> map = new HashMap<>();
+
+                    @Override
+                    protected Map<String, List<AnnotatedTypeMirror>> defaultAction(
+                            AnnotatedTypeMirror type, Void aVoid) {
+                        if (type == null) {
+                            return map;
+                        }
+                        String simple = type.toString();
+                        List<AnnotatedTypeMirror> list;
```

Comment:
Is there a library that would abstract out the process of adding to a multimap?  (Maybe not one that we wish the CF to depend upon?)

I think we can use the same visitor for both ATMs, and that doing so will not only shorten the code but will solve a problem with the current design.

It is a problem if any two types anywhere in the message are printed the same but different under the hood.  The current pull request uses verbose printing if two types in different ATMs have that property, but not if two types in the same ATM have that property.  I think it would be simpler and safer to use the same visitor for both ATMs.  It would always check for duplicates (causing `shouldPrintVerbose` to return true if there is any duplicate) and always add to the map if there isn't a duplicate.